db/compaction_job.cc

```diff
@@ -171,12 +171,18 @@ struct CompactionJob::SubcompactionState {
     const std::vector<FileMetaData*>& grandparents = compaction->grandparents();
 
     // Scan to find earliest grandparent file that contains key.
+    int grandparent_index_moved = 0;
     while (grandparent_index < grandparents.size() &&
            icmp->Compare(internal_key,
                          grandparents[grandparent_index]->largest.Encode()) >
                0) {
       if (seen_key) {
         overlapped_bytes += grandparents[grandparent_index]->fd.GetFileSize();
+        if (grandparent_index == 0 ||
+            grandparents[grandparent_index]->fd.GetFileSize() >
+                compaction->max_output_file_size() / 8) {
```

Comment:
grandparent_index_moved is set to 0 in line 174, so only if the parent index moves twice when we add a key we skip it. We can only move the index twice if the next file is totally between the current key and previous key.